pkg/controller/elasticsearch/driver/upgrade.go

```diff
@@ -194,7 +202,16 @@ func podsToUpgrade(
 				// Pod does not exist, continue the loop as the absence will be accounted by the deletion driver
 				continue
 			}
-			if sset.PodRevision(pod) != statefulSet.Status.UpdateRevision {
+			// We consider that a Pod needs an update if at least one the following condition is met:
+			// 1. The update revision of the Pod does not match the one in the status of the StatefulSet
+			// 2. The Elasticsearch version run by the Pod does not match the expected one in the Elasticsearch object
+			// Using the Pod revision might not be enough since it is not propagated consistently across all the StatefulSets.
+			// See https://github.com/elastic/cloud-on-k8s/issues/2393#issuecomment-572951884
+			podVersion, err := label.ExtractVersion(pod.Labels)
+			if err != nil {
+				return toUpgrade, err
+			}
+			if sset.PodRevision(pod) != statefulSet.Status.UpdateRevision || !podVersion.IsSame(*esVersion) {
```

Comment:
I am just curious to know whether we can be certain that the version label would be up-to-date when the revision label isn't? Is it due to a difference in timing between the the Statefulset controller and ECK? (i.e. StatefuleSet controller is more relaxed about updating the revision than ECK is about updating the version label)

Reply:
This is the revision in the status that seems to not be up to date. I didn't have the time to investigate why it can take some time to update it while the generation has been updated.